diff --git a/.mailmap b/.mailmap
index 6e138e7a8b..66b78fb919 100644
--- a/.mailmap
+++ b/.mailmap
@@ -318,6 +318,7 @@ Fabien Tassin <tassin@eerie.fr> Fabien TASSIN <tassin@eerie.fr>
 Father Chrysostomos <sprout@cpan.org> Father Chrysostomos (via RT) <perlbug-followup@perl.org>
 Father Chrysostomos <sprout@cpan.org> Father Chrysostomos <perlbug-followup@perl.org>
 François Perrad <francois.perrad@gadz.org> Francois Perrad <francois.perrad@gadz.org>
+Gavin Hayes <gavin@computoid.com> Gavin Hayes <gavin@computoid.com>
 Gene Sullivan <genesullivan50@yahoo.com> gene sullivan <genesullivan50@yahoo.com>
 Gene Sullivan <genesullivan50@yahoo.com> Gene Sullivan <gsullivan@cpan.org>
 Gene Sullivan <genesullivan50@yahoo.com> Gene Sullivan <perlbug-followup@perl.org>
diff --git a/AUTHORS b/AUTHORS
index ce9bc9d315..124dd0dc64 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -509,6 +509,7 @@ Garry T. Williams              <garry@zvolve.com>
 Gary Clark                     <GaryC@mail.jeld-wen.com>
 Gary L. Armstrong
 Gary Ng                        <71564.1743@compuserve.com>
+Gavin Hayes                    <gavin@computoid.com>
 Gavin Shelley                  <columbusmonkey@me.com>
 Gene Sullivan                  <genesullivan50@yahoo.com>
 Geoffrey F. Green              <geoff-public@stuebegreen.com>
diff --git a/Configure b/Configure
index d82bcc3ce3..d93b62f8c4 100755
--- a/Configure
+++ b/Configure
@@ -2773,7 +2773,7 @@ int main(int argc, char *argv[]) {
   return 0;
 }
 EOM
-    if $cc -o try $ccflags $ldflags try.c; then
+    if $cc -o try $ccflags $ldflags try.c $libs; then
        :
     else
         echo "Uh-oh, the C compiler '$cc' doesn't seem to be working." >&4
@@ -2845,7 +2845,7 @@ int main(int argc, char *argv[]) {
   return 0;
 }
 EOM
-    if $cc -o try $ccflags $ldflags try.c; then
+    if $cc -o try $ccflags $ldflags try.c $libs; then
        :
     else
         if $test X"$despair" = Xyes; then
@@ -3697,6 +3697,11 @@ EOM
 		fi
 	fi
 
+	if [ $COSMOCC ]; then
+		osname='cosmo'
+		osvers=''
+	fi
+
 	case "$targetarch" in
 	'') ;;
 	*)  hostarch=$osname
@@ -4608,7 +4613,7 @@ int main() {
 	return(0);
 }
 EOM
-if $cc -o try $ccflags $ldflags try.c; then
+if $cc -o try $ccflags $ldflags try.c $libs; then
 	gccversion=`$run ./try`
 	case "$gccversion" in
 	'') echo "You are not using GNU cc." ;;
@@ -4729,7 +4734,7 @@ int main(int argc, char **argv) {
 EOCP
 c99_for=no
 for flag in '' '-std=gnu99' '-std=c99'; do
-    if $cc -o try $flag $ccflags $ldflags try.c 2>/dev/null && ./try; then
+    if $cc -o try $flag $ccflags $ldflags try.c $libs 2>/dev/null && ./try; then
         c99_for="$flag"
         break;
     fi
@@ -5420,7 +5425,7 @@ echo " ";
 echo "Checking if your compiler accepts $flag" >&4;
 [ "X$sysroot" != "X" ] && echo "For sysroot = $sysroot";
 echo "int main(void) { return 0; }" > gcctest.c;
-if $cc $_sysroot -O2 $flag -o gcctest gcctest.c 2>gcctest_raw.out && $run ./gcctest; then
+if $cc $_sysroot -O2 $flag -o gcctest $ccflags $ldflags gcctest.c $libs 2>gcctest_raw.out && $run ./gcctest; then
     echo "Yes, it does." >&4;
     grep -v "when in C++ mode, this behavior is deprecated" gcctest_raw.out > gcctest.out;
     if $test -s gcctest.out ; then
@@ -24138,6 +24143,8 @@ else
 	if $test -s ccsym.own; then
 		$test "$also" && echo " "
 		echo "Your C compiler ${also}defines the following cpp symbols:"
+		echo "LARGEFILE_SOURCE" > ccsym.own
+		echo "DONT_USE_BUILTIN_SETJMP" >> ccsym.own
 		$sed -e 's/\(..*\)=1/\1/' ccsym.own
 		$sed -e 's/\(..*\)=.*/\1/' ccsym.own | $uniq >>Cppsym.true
 		ccsymbols=`$cat ccsym.own`
diff --git a/MANIFEST b/MANIFEST
index 32eb977dc2..48d4e557d2 100644
--- a/MANIFEST
+++ b/MANIFEST
@@ -138,6 +138,7 @@ README.amiga				Perl notes for AmigaOS
 README.android				Perl notes for Android
 README.bs2000				Perl notes for POSIX-BC BS2000
 README.cn				Perl for Simplified Chinese (in UTF-8)
+README.cosmo			Perl notes for the Cosmopolitan Libc
 README.cygwin				Perl notes for Cygwin
 README.freebsd				Perl notes for FreeBSD
 README.haiku				Perl notes for Haiku
@@ -5299,6 +5300,7 @@ hints/broken-db.msg			Warning message for systems with broken DB library
 hints/bsdos.sh				Hints for named architecture
 hints/catamount.sh			Hints for named architecture
 hints/convexos.sh			Hints for named architecture
+hints/cosmo.sh				Hints for named architecture
 hints/cxux.sh				Hints for named architecture
 hints/cygwin.sh				Hints for named architecture
 hints/darwin.sh				Hints for named architecture
diff --git a/Makefile.SH b/Makefile.SH
index 6f1a114640..9da47e756d 100755
--- a/Makefile.SH
+++ b/Makefile.SH
@@ -1017,6 +1017,19 @@ lib/buildcustomize.pl: $& $(miniperl_objs) write_buildcustomize.pl
 	    $(miniperl_objs) $(libs)
 	$(LDLIBPTH) ./miniperl$(HOST_EXE_EXT) -w -Ilib -Idist/Exporter/lib -MExporter -e '<?>' || sh -c 'echo >&2 Failed to build miniperl.  Please run make minitest; exit 1'
 	$(MINIPERL) -f write_buildcustomize.pl
+!NO!SUBS!
+		;;
+	cosmo*)
+		$spitshell >>$Makefile <<'!NO!SUBS!'
+lib/buildcustomize.pl: $& $(miniperl_dep) write_buildcustomize.pl
+	-@rm -f miniperl.xok
+	$(CC) $(CLDFLAGS) -o miniperl.com \
+	    $(miniperl_objs) $(libs)
+	cp miniperl.com $(MINIPERL_EXE)
+	chmod -w miniperl.com
+	./miniperl --assimilate || true
+	$(LDLIBPTH) ./miniperl$(HOST_EXE_EXT) -w -Ilib -Idist/Exporter/lib -MExporter -e '<?>' || sh -c 'echo >&2 Failed to build miniperl.  Please run make minitest; exit 1'
+	$(MINIPERL) -f write_buildcustomize.pl
 !NO!SUBS!
 		;;
 	*)
@@ -1072,7 +1085,13 @@ $(PERL_EXE): $& $(perlmain_dep) $(LIBPERL) $(static_ext) ext.libs $(PERLEXPORT)
 	       ;;
 	    esac
         ;;
-
+        cosmo) $spitshell >>$Makefile <<'!NO!SUBS!'
+	$(SHRPENV) $(CC) -o perl.com $(CLDFLAGS) $(CCDLFLAGS) $(perlmain_objs) $(static_ext) $(LLIBPERL) `cat ext.libs` $(libs)
+	cp perl.com $(PERL_EXE)
+	chmod -w perl.com
+	./perl --assimilate || true
+!NO!SUBS!
+        ;;
         *) $spitshell >>$Makefile <<'!NO!SUBS!'
 	$(SHRPENV) $(CC) -o perl $(CLDFLAGS) $(CCDLFLAGS) $(perlmain_objs) $(static_ext) $(LLIBPERL) `cat ext.libs` $(libs)
 !NO!SUBS!
diff --git a/Porting/Maintainers.pl b/Porting/Maintainers.pl
index 2d428d76e8..33decf99a1 100755
--- a/Porting/Maintainers.pl
+++ b/Porting/Maintainers.pl
@@ -1391,6 +1391,7 @@ our %Modules = (
                 ext/Devel-Peek/
                 ext/DynaLoader/
                 ext/Errno/
+                ext/ErrnoRuntime/
                 ext/ExtUtils-Miniperl/
                 ext/Fcntl/
                 ext/File-DosGlob/
diff --git a/README.cosmo b/README.cosmo
new file mode 100644
index 0000000000..e5b5de8952
--- /dev/null
+++ b/README.cosmo
@@ -0,0 +1,87 @@
+If you read this file _as_is_, just ignore the funny characters you see.
+It is written in the POD format (see pod/perlpod.pod) which is specially
+designed to be readable as is.
+
+=head1 NAME
+
+perlcosmo - Build-once run-anywhere Perl using the Cosmopolitan Libc
+
+=head1 DESCRIPTION
+
+The cosmo platform addition eases building Perl with the Cosmopolitan
+Libc. When you build with the Cosmopolitan Libc, the libc is the
+platform as the Cosmopolitan Libc targets building simultaneous support
+for multiple operating systems (Unixes, Windows, etc) into the same
+binary.
+
+To get the most benefits of the cosmo platform, you likely want to
+build Actually Portable Perl (APPerl) instead of this directly. See
+L<Perl::Dist::APPerl> ; APPerl Manager (apperlm) includes a front-end
+for building perl from source with the Cosmopolitan Libc. The
+instructions in this document only cover building perl, not building or
+packaging into APPerl (single executable perl with perl modules
+embedded inside).
+
+=head1 BUILD
+
+Download the Cosmopolitan Libc and build with the desired MODE:
+
+  git clone https://github.com/jart/cosmopolitan ../cosmopolitan
+  make -C ../cosmopolitan -j4 MODE=
+
+Run Configure, but with some environment variables to point it at
+cosmopolitan; update COSMO_MODE to the MODE you set before and
+COSMO_REPO to the absolute path of the cosmopolitan repo. Then, make
+and install like usual.
+
+  COSMO_MODE= COSMO_REPO="/home/gavin/cosmopolitan" ./Configure -de
+  make
+  make install
+
+Note the installed binary is assimilated (see Cosmopolitan Libc
+--assimilate flag), meaning it only runs on the current operating
+system. The perl.com in the perl repo, however, is actually portable
+and will run on several operating systems. For a solution that builds
+on this, by packaging perl modules and scripts inside of the executable
+and more, see L<Perl::Dist::APPerl>.
+
+=head1 KNOWN PROBLEMS
+
+cosmo perl is a work in progress, many major issues are present.
+
+=over 4
+
+=item *
+
+C<system> behavior when a shell is required, depends on the currently
+running operating system. On most systems it uses, C<sh>, but on
+Windows it uses C<cmd.exe>. Therefore, it's still hard to use portably.
+However, it is sufficient for pulling up a pager for L<perldoc> .
+
+=item *
+
+After building, only Pure Perl modules may be added as perl is built
+statically.
+
+=item *
+
+Some magic only works when the binary is assimilated. For example,
+modifying C<$0>. Running with C<--assimilate> will bind the binary to
+the current operating system.
+
+=back
+
+=head1 ACKNOWLEDGEMENTS
+
+The L<Cosmopolitan Libc|https://github.com/jart/cosmopolitan>
+contributors, especially L<Justine Tunney|https://justine.lol/> and
+L<Gautham Venkatasubramanian|https://ahgamut.github.io>.
+Gautham's
+L<Python port|https://ahgamut.github.io/2021/07/13/ape-python/>
+to the Cosmopolitan Libc inspired this port.
+
+=head1 CONTACT
+
+The port in-progress is being done by Gavin Hayes <gahayes@cpan.org>
+
+Last update: 2022-10-11
diff --git a/caretx.c b/caretx.c
index 855608b5e0..a1805117f9 100644
--- a/caretx.c
+++ b/caretx.c
@@ -40,6 +40,10 @@
 # include <mach-o/dyld.h>
 #endif
 
+#if defined(__COSMOPOLITAN__)
+#  include "libc/cosmo.h"
+#endif
+
 void
 Perl_set_caret_X(pTHX)
 {
@@ -58,6 +62,9 @@ Perl_set_caret_X(pTHX)
     sv_setpv(caret_x, ansi);
     win32_free(ansi);
     return;
+#elif defined(__COSMOPOLITAN__)
+    sv_setpv(caret_x, GetProgramExecutableName());
+    return;
 #else
     /* We can try a platform-specific one if possible; if it fails, or we
      * aren't running on a suitable platform, we'll fall back to argv[0]. */
diff --git a/configpm b/configpm
index 0c46b9e6da..3ba9bcf527 100755
--- a/configpm
+++ b/configpm
@@ -592,6 +592,13 @@ $heavy_txt .= join('',
     } (@v_others, @v_forced, @v_define)
 ) . "!END!\n";
 
+if ($^O eq 'cosmo') {
+$heavy_txt .= <<'EOT';
+
+s/perlpath=.+/perlpath='$^X'/;
+EOT
+}
+
 # Only need the dynamic byteorder code in Config.pm if 'byteorder' is one of
 # the precached keys
 if ($Common{byteorder}) {
diff --git a/cpan/Archive-Tar/t/04_resolved_issues.t b/cpan/Archive-Tar/t/04_resolved_issues.t
index 08d339ac87..69345076ae 100644
--- a/cpan/Archive-Tar/t/04_resolved_issues.t
+++ b/cpan/Archive-Tar/t/04_resolved_issues.t
@@ -275,6 +275,11 @@ SKIP: {
   ok( 1,                      "Testing bug 103279" );
 	my $tar = $Class->new;
 	isa_ok( $tar, $Class,       "   Object" );
+	if ($^O eq 'cosmo') {
+		# Cosmopolitan prevents creating files that end with space so behavior
+		# doesn't change between running on Windows and elsewhere.
+		skip 'Cosmo doesn\'t allow creating files that end with space', 7;
+	}
 	ok( open my $fh, '>', 'white_space   ' );
 	ok( close $fh );
 	if (-e 'white_space' && $^O eq 'MSWin32') {
diff --git a/cpan/ExtUtils-Constant/lib/ExtUtils/Constant/ProxySubs.pm b/cpan/ExtUtils-Constant/lib/ExtUtils/Constant/ProxySubs.pm
index 0aee5233fe..cd17c1c86f 100644
--- a/cpan/ExtUtils-Constant/lib/ExtUtils/Constant/ProxySubs.pm
+++ b/cpan/ExtUtils-Constant/lib/ExtUtils/Constant/ProxySubs.pm
@@ -406,13 +406,14 @@ EOC
 
 	print $c_fh "struct $struct_type $struct;\n";
 
+    my $structtypetype = $type ? 'static struct' : 'static const struct';
 	print $struct_fh <<"EOBOOT";
 
-    static const struct $struct_type $array_name\[] =
+    $structtypetype $struct_type $array_name\[] =
       {
 EOBOOT
 
-
+    my @valuestowrite;
 	foreach my $item (@{$found->{$type}}) {
             my ($name, $namelen, $value, $macro)
                  = $self->name_len_value_macro($item);
@@ -429,8 +430,8 @@ EOBOOT
 	    } else {
 		print $struct_fh $ifdef;
 	    }
-	    print $struct_fh "        { ", join (', ', "\"$name\"", $namelen,
-						 &$type_to_value($value)),
+		push @valuestowrite, &$type_to_value($value);
+	    print $struct_fh "        { ", join (', ', "\"$name\"", $namelen),
 						 " },\n",
 						 $self->macro_to_endif($macro);
 	}
@@ -438,6 +439,28 @@ EOBOOT
     # Terminate the list with a NULL
 	print $struct_fh "        { NULL, 0", (", 0" x $number_of_args), " } };\n";
 
+	if($type) {
+		print $struct_fh "{\nunsigned i = 0;\n";
+		foreach my $item (@{$found->{$type}}) {
+            my ($name, $namelen, $value, $macro) = $self->name_len_value_macro($item);
+	        my $ifdef = $self->macro_to_ifdef($macro);
+	        if (!$ifdef && $item->{invert_macro}) {
+	        carp("Attempting to supply a default for '$name' which has no conditional macro");
+	        next;
+	        }
+	        if ($item->{invert_macro}) {
+	        print $struct_fh $self->macro_to_ifndef($macro);
+	        print $struct_fh
+	        "        /* This is the default value: */\n" if $type;
+	        } else {
+	        print $struct_fh $ifdef;
+	        }
+			print $struct_fh $array_name,"[i++].value = ". (shift @valuestowrite) . ";\n";
+			print $struct_fh $self->macro_to_endif($macro);
+	    }
+	    print $struct_fh "}\n";
+	}
+
 	print $xs_fh <<"EOBOOT" if $type;
 	const struct $struct_type *$iterator{$type} = $array_name;
 EOBOOT
diff --git a/cpan/File-Path/t/Path.t b/cpan/File-Path/t/Path.t
index b265aeeb2d..e9812c5f69 100644
--- a/cpan/File-Path/t/Path.t
+++ b/cpan/File-Path/t/Path.t
@@ -370,7 +370,7 @@ SKIP: {
   # Better to search for *nix derivatives?
   # Not sure what else doesn't support newline in paths
   skip "$^O doesn't allow newline in paths", 2
-    if $^O =~ m/^(MSWin32|VMS)$/;
+    if $^O =~ m/^(MSWin32|VMS|cosmo)$/;
 
   @created = make_path( $dir, $dir2 );
 
diff --git a/cpan/Perl-OSType/lib/Perl/OSType.pm b/cpan/Perl-OSType/lib/Perl/OSType.pm
index a72dd024b5..a0f2606137 100644
--- a/cpan/Perl-OSType/lib/Perl/OSType.pm
+++ b/cpan/Perl-OSType/lib/Perl/OSType.pm
@@ -55,6 +55,7 @@ my %OSTYPES = qw(
   nto         Unix
   qnx         Unix
   android     Unix
+  cosmo       Unix
 
   dos         Windows
   MSWin32     Windows
diff --git a/dist/IO/t/cachepropagate-unix.t b/dist/IO/t/cachepropagate-unix.t
index 718d01da1d..8024a10474 100644
--- a/dist/IO/t/cachepropagate-unix.t
+++ b/dist/IO/t/cachepropagate-unix.t
@@ -50,7 +50,7 @@ my $p = $listener->protocol();
     # This is a TODO instead of a skip so if these ever implement SO_PROTOCOL
     # we'll be notified about the passing TODO so the test can be updated.
     local $TODO = "$^O doesn't support SO_PROTOCOL on AF_UNIX"
-        if $^O =~ /^(netbsd|darwin|cygwin|hpux|solaris|dragonfly|os390|gnu)$/;
+        if $^O =~ /^(netbsd|darwin|cygwin|hpux|solaris|dragonfly|os390|gnu|cosmo)$/;
     ok(defined($p), 'protocol defined');
 }
 my $d = $listener->sockdomain();
@@ -105,7 +105,7 @@ SKIP: {
     {
         # see comment above
         local $TODO = "$^O doesn't support SO_PROTOCOL on AF_UNIX"
-            if $^O =~ /^(netbsd|darwin|cygwin|hpux|solaris|dragonfly|os390|gnu)$/;
+            if $^O =~ /^(netbsd|darwin|cygwin|hpux|solaris|dragonfly|os390|gnu|cosmo)$/;
         ok(defined($p), 'protocol defined');
     }
     $d = $listener->sockdomain();
diff --git a/dist/Time-HiRes/Makefile.PL b/dist/Time-HiRes/Makefile.PL
index ac56d8df27..3fe55c007e 100644
--- a/dist/Time-HiRes/Makefile.PL
+++ b/dist/Time-HiRes/Makefile.PL
@@ -90,7 +90,7 @@ __EOD__
             $errornull = '';
         }
 
-        $cccmd = "$Config{'cc'} -o $tmp $ccflags $tmp.c @$LIBS $errornull"
+        $cccmd = "$Config{'cc'} -o $tmp $ccflags $Config{'ldflags'} $tmp.c @$LIBS $Config{'libs'} $errornull"
             unless defined $cccmd;
 
        if ($^O eq 'VMS') {
diff --git a/doio.c b/doio.c
index 87b9e8498d..54c497a8bc 100644
--- a/doio.c
+++ b/doio.c
@@ -55,6 +55,11 @@
 #  define OPEN_EXCL 0
 #endif
 
+#ifdef __COSMOPOLITAN__
+#    include "libc/dce.h"
+#    include "libc/runtime/runtime.h"
+#endif
+
 #define PERL_MODE_MAX 8
 #define PERL_FLAGS_MAX 10
 
@@ -2550,7 +2555,24 @@ Perl_do_exec3(pTHX_ const char *incmd, int fd, int do_report)
             unsetenv("_TAG_REDIR_ERR");
   #endif
 #endif
-            PerlProc_execl(PL_sh_path, "sh", "-c", cmd, (char *)NULL);
+#ifdef __COSMOPOLITAN__
+            if(IsWindows())
+            {
+                // hack for perldoc
+                static const char paginate_cmd_start[] = "more \"/tmp/";
+                if (strncmp(paginate_cmd_start, cmd, sizeof(paginate_cmd_start)-1) == 0) {
+                    cmd[strlen(cmd)-1] = '\0';
+                    __paginate_file(1, cmd + 6);
+                    exit(0);
+                }
+
+                PerlProc_execl("/C/Windows/System32/cmd.exe", "/C/Windows/System32/cmd.exe", "/c", cmd, (char *)NULL);
+            }
+            else
+#endif
+            {
+                PerlProc_execl(PL_sh_path, "sh", "-c", cmd, (char *)NULL);
+            }
             PERL_FPU_POST_EXEC
             S_exec_failed(aTHX_ PL_sh_path, fd, do_report);
             goto leave;
diff --git a/ext/DynaLoader/t/DynaLoader.t b/ext/DynaLoader/t/DynaLoader.t
index 11b37b5c4e..7cd7769d38 100644
--- a/ext/DynaLoader/t/DynaLoader.t
+++ b/ext/DynaLoader/t/DynaLoader.t
@@ -119,7 +119,7 @@ SKIP: {
     # (not at least by that name) that the dl_findfile()
     # could find.
     skip( "dl_findfile test not appropriate on $^O", 1 )
-	if $^O =~ /(win32|vms|openbsd|bitrig|cygwin|vos|os390)/i;
+	if $^O =~ /(win32|vms|openbsd|bitrig|cygwin|vos|os390|cosmo)/i;
     # Play safe and only try this test if this system
     # looks pretty much Unix-like.
     skip( "dl_findfile test not appropriate on $^O", 1 )
diff --git a/ext/POSIX/t/sigaction.t b/ext/POSIX/t/sigaction.t
index 024c66a147..2741c895ed 100644
--- a/ext/POSIX/t/sigaction.t
+++ b/ext/POSIX/t/sigaction.t
@@ -51,7 +51,7 @@ ok($oldaction->{MASK}->ismember(SIGUSR1), "SIGUSR1 ismember MASK");
 
 SKIP: {
     skip("sigaction() thinks different in $^O", 1)
-	if $^O eq 'linux' || $^O eq 'unicos';
+	if $^O eq 'linux' || $^O eq 'unicos' || $^O eq 'cosmo';
     is($oldaction->{FLAGS}, 0);
 }
 
diff --git a/ext/POSIX/t/sigset.t b/ext/POSIX/t/sigset.t
index 864b4f528e..4d903b273e 100644
--- a/ext/POSIX/t/sigset.t
+++ b/ext/POSIX/t/sigset.t
@@ -34,6 +34,7 @@ sub expected_signals {
     }
 }
 
+my @fillsetsigno;
 foreach (@POSIX::EXPORT) {
     next unless /^SIG[A-Z0-9]+$/;
     my $val = eval "POSIX::$_";
@@ -41,6 +42,9 @@ foreach (@POSIX::EXPORT) {
     $min = $val if $min > $val;
     $max = $val if $max < $val;
     push @signo, $val;
+     # Cosmpolitian doesn't allow adding these signals with sigfillset
+    next if $^O eq 'cosmo' && /^(?:SIGTHR|SIGABRT|SIGKILL|SIGSTOP)$/;
+    push @fillsetsigno, $val;
 }
 
 # Sanity check that we found something:
@@ -55,7 +59,7 @@ is($sigset->fillset(), '0 but true', 'fillset');
 # because on some systems, not all integers are valid signals...
 # so the only thing we can really be confident about is that all the signals
 # with names are going to be present:
-foreach (@signo) {
+foreach (@fillsetsigno) {
     cmp_ok($sigset->ismember($_), '==', 1, "after fillset sig $_ is a member");
 }
 is($sigset->emptyset(), '0 but true', 'empyset');
diff --git a/hints/cosmo.sh b/hints/cosmo.sh
new file mode 100644
index 0000000000..72058dddbf
--- /dev/null
+++ b/hints/cosmo.sh
@@ -0,0 +1,22 @@
+#! /bin/sh
+# cosmo.sh - hints for building perl using the Cosmopolitan Libc
+#
+
+test -d "$COSMOCC" || exit 1;
+
+usedl='undef'
+usenm='false'
+so='none'
+osname='cosmo'
+osvers=''
+libpth=''
+d_procselfexe='undef'
+i_stdckdint='undef'
+locincpth=''
+loclibpth=''
+glibpth=''
+cc="$COSMOCC/bin/x86_64-unknown-cosmo-cc"
+ccflags="-fno-stack-protector -D_COSMO_SOURCE"
+ldflags="-fno-stack-protector"
+libs=''
+unset COSMOCC
diff --git a/lib/.gitignore b/lib/.gitignore
index 2107e177ec..899607478f 100644
--- a/lib/.gitignore
+++ b/lib/.gitignore
@@ -37,6 +37,7 @@
 /Encode/
 /Env.pm
 /Errno.pm
+/ErrnoRuntime.pm
 /Exporter.pm
 /Exporter/
 /ExtUtils/CBuilder.pm
diff --git a/patchlevel.h b/patchlevel.h
index 584d749c78..8073fc0867 100644
--- a/patchlevel.h
+++ b/patchlevel.h
@@ -156,6 +156,7 @@ hunk.
 static const char * const local_patches[] = {
         NULL
 		, "RC2"
+		, "Perl-Dist-APPerl-v0.8.0/5.44-cosmo3.patch - https://github.com/G4Vi/Perl-Dist-APPerl/blob/v0.8.0/share/5.44-cosmo3.patch various changes to support building with the Cosmopolitan Libc"
 #ifdef PERL_GIT_UNCOMMITTED_CHANGES
         ,"uncommitted-changes"
 #endif
diff --git a/pod/.gitignore b/pod/.gitignore
index 264f4f59a7..da6a82a0b9 100644
--- a/pod/.gitignore
+++ b/pod/.gitignore
@@ -4,6 +4,7 @@
 /perlandroid.pod
 /perlbs2000.pod
 /perlcn.pod
+/perlcosmo.pod
 /perlcygwin.pod
 /perldos.pod
 /perlfreebsd.pod
diff --git a/pod/perl.pod b/pod/perl.pod
index eee74c0a80..16c765ab0d 100644
--- a/pod/perl.pod
+++ b/pod/perl.pod
@@ -295,6 +295,7 @@ aux h2ph h2xs perlbug pl2pm pod2html pod2man splain xsubpp
     perlamiga           Perl notes for AmigaOS
     perlandroid         Perl notes for Android
     perlbs2000          Perl notes for POSIX-BC BS2000
+    perlcosmo           Perl notes for the Cosmopolitan Libc
     perlcygwin          Perl notes for Cygwin
     perlfreebsd         Perl notes for FreeBSD
     perlhaiku           Perl notes for Haiku
diff --git a/t/op/magic.t b/t/op/magic.t
index 929a746742..0218a8b233 100644
--- a/t/op/magic.t
+++ b/t/op/magic.t
@@ -277,7 +277,7 @@ $$ = $pid; # Tests below use $$
 # $^X and $0
 {
     my $is_abs = $Config{d_procselfexe} || $Config{usekernprocpathname}
-      || $Config{usensgetexecutablepath};
+      || $Config{usensgetexecutablepath} || ($^O eq 'cosmo');
     if ($^O eq 'qnx') {
 	chomp($wd = `/usr/bin/fullpath -t`);
     }
diff --git a/t/porting/known_pod_issues.dat b/t/porting/known_pod_issues.dat
index 3f55205702..951e778cc5 100644
--- a/t/porting/known_pod_issues.dat
+++ b/t/porting/known_pod_issues.dat
@@ -1,4 +1,4 @@
-# This file is the data file for porting/podcheck.t.
+# This file is the data file for t/porting/podcheck.t.
 # There are three types of lines.
 # Comment lines are white-space only or begin with a '#', like this one.  Any
 #   changes you make to the comment lines will be lost when the file is
@@ -262,6 +262,7 @@ pclose(3)
 perl(1)
 Perl4::CoreLibs
 perl5198delta
+Perl::Dist::APPerl
 Perl::Unsafe::Signals
 perlbug(1)
 PerlIO::locale
diff --git a/win32/GNUmakefile b/win32/GNUmakefile
index 3c3291d283..a6a46a1122 100644
--- a/win32/GNUmakefile
+++ b/win32/GNUmakefile
@@ -1624,6 +1624,7 @@ utils: $(HAVEMINIPERL) ..\utils\Makefile
 	copy ..\README.android  ..\pod\perlandroid.pod
 	copy ..\README.bs2000   ..\pod\perlbs2000.pod
 	copy ..\README.cn       ..\pod\perlcn.pod
+	copy ..\README.cosmo    ..\pod\perlcosmo.pod
 	copy ..\README.cygwin   ..\pod\perlcygwin.pod
 	copy ..\README.freebsd  ..\pod\perlfreebsd.pod
 	copy ..\README.haiku    ..\pod\perlhaiku.pod
@@ -1748,14 +1749,14 @@ distclean: realclean
 	-if exist $(LIBDIR)\XS rmdir /s /q $(LIBDIR)\XS
 	-cd $(PODDIR) && del /f *.html *.bat roffitall \
 	    perl5440delta.pod perlaix.pod perlamiga.pod perlandroid.pod \
-	    perlapi.pod perlbs2000.pod perlcn.pod perlcygwin.pod \
-	    perlfreebsd.pod perlhaiku.pod perlhpux.pod perlhurd.pod \
-	    perlintern.pod perlirix.pod perljp.pod perlko.pod perllinux.pod \
-	    perlmacosx.pod perlmodlib.pod perlopenbsd.pod perlos2.pod \
-	    perlos390.pod perlos400.pod perlplan9.pod perlqnx.pod \
-	    perlriscos.pod perlsolaris.pod perlsynology.pod perltoc.pod \
-	    perltru64.pod perltw.pod perluniprops.pod perlvos.pod \
-	    perlwin32.pod
+	    perlapi.pod perlbs2000.pod perlcn.pod perlcosmo.pod \
+	    perlcygwin.pod perlfreebsd.pod perlhaiku.pod perlhpux.pod \
+	    perlhurd.pod perlintern.pod perlirix.pod perljp.pod perlko.pod \
+	    perllinux.pod perlmacosx.pod perlmodlib.pod perlopenbsd.pod \
+	    perlos2.pod perlos390.pod perlos400.pod perlplan9.pod \
+	    perlqnx.pod perlriscos.pod perlsolaris.pod perlsynology.pod \
+	    perltoc.pod perltru64.pod perltw.pod perluniprops.pod \
+	    perlvos.pod perlwin32.pod
 	-cd ..\utils && del /f h2ph splain perlbug pl2pm h2xs \
 	    perldoc perlivp libnetcfg enc2xs encguess piconv cpan *.bat \
 	    xsubpp pod2html instmodsh json_pp prove ptar ptardiff ptargrep shasum corelist zipdetails
diff --git a/win32/Makefile b/win32/Makefile
index 42d452e19c..17c1304014 100644
--- a/win32/Makefile
+++ b/win32/Makefile
@@ -1151,6 +1151,7 @@ utils: $(PERLEXE) ..\utils\Makefile
 	copy ..\README.android  ..\pod\perlandroid.pod
 	copy ..\README.bs2000   ..\pod\perlbs2000.pod
 	copy ..\README.cn       ..\pod\perlcn.pod
+	copy ..\README.cosmo    ..\pod\perlcosmo.pod
 	copy ..\README.cygwin   ..\pod\perlcygwin.pod
 	copy ..\README.freebsd  ..\pod\perlfreebsd.pod
 	copy ..\README.haiku    ..\pod\perlhaiku.pod
@@ -1276,14 +1277,14 @@ distclean: realclean
 	-if exist $(LIBDIR)\XS rmdir /s /q $(LIBDIR)\XS
 	-cd $(PODDIR) && del /f *.html *.bat roffitall \
 	    perl5440delta.pod perlaix.pod perlamiga.pod perlandroid.pod \
-	    perlapi.pod perlbs2000.pod perlcn.pod perlcygwin.pod \
-	    perlfreebsd.pod perlhaiku.pod perlhpux.pod perlhurd.pod \
-	    perlintern.pod perlirix.pod perljp.pod perlko.pod perllinux.pod \
-	    perlmacosx.pod perlmodlib.pod perlopenbsd.pod perlos2.pod \
-	    perlos390.pod perlos400.pod perlplan9.pod perlqnx.pod \
-	    perlriscos.pod perlsolaris.pod perlsynology.pod perltoc.pod \
-	    perltru64.pod perltw.pod perluniprops.pod perlvos.pod \
-	    perlwin32.pod
+	    perlapi.pod perlbs2000.pod perlcn.pod perlcosmo.pod \
+	    perlcygwin.pod perlfreebsd.pod perlhaiku.pod perlhpux.pod \
+	    perlhurd.pod perlintern.pod perlirix.pod perljp.pod perlko.pod \
+	    perllinux.pod perlmacosx.pod perlmodlib.pod perlopenbsd.pod \
+	    perlos2.pod perlos390.pod perlos400.pod perlplan9.pod \
+	    perlqnx.pod perlriscos.pod perlsolaris.pod perlsynology.pod \
+	    perltoc.pod perltru64.pod perltw.pod perluniprops.pod \
+	    perlvos.pod perlwin32.pod
 	-cd ..\utils && del /f h2ph splain perlbug pl2pm h2xs \
 	    perldoc perlivp libnetcfg enc2xs encguess piconv cpan streamzip *.bat \
 	    xsubpp pod2html instmodsh json_pp prove ptar ptardiff ptargrep shasum corelist zipdetails
